test/integration/fake_upstream.h

```diff
@@ -239,33 +251,44 @@ using FakeStreamPtr = std::unique_ptr<FakeStream>;
 // SharedConnectionWrapper that lives from when the Connection is added to the accepted connection
 // queue and then through the lifetime of the Fake{Raw,Http}Connection that manages the Connection
 // through active use.
-class SharedConnectionWrapper : public Network::ConnectionCallbacks {
+class SharedConnectionWrapper : public Network::ConnectionCallbacks,
+                                public LinkedObject<SharedConnectionWrapper> {
 public:
   using DisconnectCallback = std::function<void()>;
 
   SharedConnectionWrapper(Network::Connection& connection, bool allow_unexpected_disconnects)
       : connection_(connection), allow_unexpected_disconnects_(allow_unexpected_disconnects) {
     connection_.addConnectionCallbacks(*this);
+    addDisconnectCallback([this] {
+      lock_.AssertReaderHeld(); // Locked in onEvent().
```

Comment:
can we drop some TODOs to move some of the multi-line impls into fake_upstream.cc?

Or better yet have some smaller files as well.

Might make compiles go faster?

Reply:
Yes I agree. I will add a TODO. This change is already pretty large. I will do a follow up.